compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -8568,24 +8568,25 @@ private BType checkArrayIndexBasedAccess(BLangIndexBasedAccess indexBasedAccess,
                 actualType = arrayType.eType;
                 break;
             case TypeTags.UNION:
-                // address the case where we have a union of finite types
-                List<BFiniteType> finiteTypes = ((BUnionType) indexExprType).getMemberTypes().stream()
-                        .filter(memType -> Types.getReferredType(memType).tag == TypeTags.FINITE)
-                        .map(matchedType -> (BFiniteType) Types.getReferredType(matchedType))
-                        .collect(Collectors.toList());
-
-                BFiniteType finiteType;
-                if (finiteTypes.size() == 1) {
-                    finiteType = finiteTypes.get(0);
-                } else {
-                    Set<BLangExpression> valueSpace = new LinkedHashSet<>();
-                    finiteTypes.forEach(constituent -> valueSpace.addAll(constituent.getValueSpace()));
-                    finiteType = new BFiniteType(null, valueSpace);
+                // address the case where we have a union of types
```

Comment:
Example 1
```ballerina
public function main() {
    record {|
        int a;
    |} r = {a: 1};
    string s = "a";
    r[s] = 2; 
}
```
Example 2
```ballerina
public function main() {
    record {|
        int a;
    |} r = {a: 1};
    string|string s = "a";
    r[s] = 2; 
}
```
When considering the first example, it is working. In the second example, we model `string|string` as a union type with only one member (which is `string`). Isn't this the actual root cause?

If it is a union type, it should contain more than one member. So here`string|string` should be equivelent to the `string` type.

Reply:
Here, I have a different opinion. The `string|string:Char` type should be modelled as a `string` type, rather than a union of those two types.

Reply:
I think we'll still need this change for a case like,
```
int[] arr = [1, 2, 3];
int:Signed16|int:Unsigned16 a = 1;
arr[a] = 4;
```

Reply:
@lochana-chathura will this fixed by your semtype integration? 
```
int[] arr = [1, 2, 3];
int:Signed16|int:Unsigned16 a = 1;
arr[a] = 4;
```
Even for this scenario also, we can represent the union type `int:Signed16|int:Unsigned16` as set containing union of both types. Is it possible with the current `BType` implementation? 

WDYT?

looping @hasithaa as well

Reply:
Yes, semtype can fix this. `BFiniteType` keeps a list of literal values. So adding all the values of `int:Signed16|int:Unsigned16` to it is not practical.  The current logic from the PR looks okay to me.